### PR TITLE
feat(authentication): Native authentication profile login

### PR DESCRIPTION
**Pull request by Calinteodor**

### CODE REVIEW

File: react/features/authentication/middleware.ts [hunk 1]
```diff
@@ -305,3 +280,36 @@ function _handleLogin({ dispatch, getState }: IStore) {
...
+
+        // user is logging out remove auto redirect indication
+        dispatch(setTokenAuthUrlSuccess(false));
+
+        if (logoutUrl) {
+            dispatch(openLogoutDialog(logoutUrl));
+        }
+    }
+
+    conference.room.moderator.logout(() => dispatch(hangup(true)));
```
Comment (Calinteodor) [inline react/features/authentication/middleware.ts:314]: @saghul Do we need to do this on both web and native?

File: lang/main.json [hunk 2]
```diff
@@ -275,6 +275,9 @@
         "alreadySharedVideoTitle": "Only one shared video is allowed at a time",
         "applicationWindow": "Application window",
         "authenticationRequired": "Authentication required",
+        "authenticationLogIn": "Log in",
```
Comment (saghul) [inline lang/main.json:278]: Don't we have all of these for web already?

File: react/features/authentication/middleware.ts (hunk 1 above)
Comment (saghul) [reply react/features/authentication/middleware.ts:314]: I think we can drop it, right @damencho ? Or won't it wipe the session id?

Comment (Calinteodor) [reply react/features/authentication/middleware.ts:314]: As a matter of fact, I think we do. I was testing with one participant and, of course, that participant is a moderator.

Comment (saghul) [reply react/features/authentication/middleware.ts:314]: Let's skip for RN for now.

File: react/features/mobile/navigation/components/HeaderNavigationButton.tsx [hunk 3]
```diff
@@ -64,6 +69,7 @@ const HeaderNavigationButton = ({ disabled, label, onPress, src, style, twoActio
             {
                 src ? (
                     <IconButton
+                        color = { src && color }
```
Comment (saghul) [inline react/features/mobile/navigation/components/HeaderNavigationButton.tsx:72]: Why is src taken taken into account here?

File: react/features/settings/actions.web.ts [hunk 4]
```diff
@@ -28,15 +31,29 @@ import {
...
+            onLogout() {
+                if (browser.isElectron()) {
+                    let url = appendURLHashParam(logoutUrl, 'electron', 'true');
+                    window.open(url, '_blank');
+                    dispatch(hangup(true));
+                }
+
+                window.location.href = logoutUrl;
+
+                return;
```
Comment (saghul) [inline react/features/settings/actions.web.ts:53]: unnecessary

File: react/features/settings/components/native/ProfileView.tsx [hunk 5]
```diff
@@ -48,6 +64,50 @@ const ProfileView = ({ isInWelcomePage }: { isInWelcomePage?: boolean; }) => {
...
+        <HeaderNavigationButton
+            color = { BaseThemeNative.palette.link01 }
+            onPress = { goBack }
+            src = { IconArrowLeft }
+            style = { styles.backBtn }
+            twoActions = { true } />
+    );
+
+    const headerRight = () => {
+        if (logoutUrl) {
```
Comment (saghul) [inline react/features/settings/components/native/ProfileView.tsx:85]: The URL will exist even before you login so looks like this logic won't work? You need to know if you are logged in or not. Also, the URL works for the token case, but the normal XMPP auth also supports logout.

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments, PTAL!

File: react/features/mobile/navigation/components/HeaderNavigationButton.tsx (hunk 3 above)
Comment (Calinteodor) [reply react/features/mobile/navigation/components/HeaderNavigationButton.tsx:72]: It's not mandatory. I can remove this. I wanted to be more specific about the use of IconButton props but there is no need to do that.

File: lang/main.json (hunk 2 above)
Comment (saghul) [reply lang/main.json:277]: Can you check what the web dialog does pl?

File: react/features/settings/components/native/ProfileView.tsx (hunk 5 above)
Comment (Calinteodor) [reply react/features/settings/components/native/ProfileView.tsx]: Updated. Used authLogin from store.

File: lang/main.json (hunk 2 above)
Comment (Calinteodor) [reply lang/main.json]: Updated.

File: react/features/authentication/middleware.ts [hunk 6]
```diff
@@ -305,3 +279,34 @@ function _handleLogin({ dispatch, getState }: IStore) {
...
+    const logoutUrl = config.tokenLogoutUrl;
+
+    const { conference } = state['features/base/conference'];
+    const { jwt } = state['features/base/jwt'];
+
+    if (!conference) {
+        return;
+    }
+
+    if (isTokenAuthEnabled(config) && config.tokenAuthUrlAutoRedirect && jwt) {
```
Comment (saghul) [inline react/features/authentication/middleware.ts:303]: Drop the redirect check pl, it only applies to web

File: react/features/settings/actions.web.ts [hunk 7]
```diff
@@ -28,15 +31,33 @@ import {
...
+
+                    window.open(url, '_blank');
+                    dispatch(hangup(true));
+                } else {
+                    window.location.href = logoutUrl;
+                }
+            }
+        }));
+
+        conference?.room.moderator.logout(() => dispatch(hangup(true)));
```
Comment (saghul) [inline react/features/settings/actions.web.ts:59]: Doesn't this need to be inside the onLogout function? Also it would likely need to run before the redirects.

File: react/features/settings/components/native/ProfileView.tsx [hunk 8]
```diff
@@ -30,6 +43,8 @@ const ProfileView = ({ isInWelcomePage }: { isInWelcomePage?: boolean; }) => {
     const [ displayName, setDisplayName ] = useState(reduxDisplayName);
     const [ email, setEmail ] = useState(reduxEmail);
 
+    const { authLogin: reduxAuthLogin } = useSelector((state: IReduxState) => state['features/base/conference']);
```
Comment (saghul) [inline react/features/settings/components/native/ProfileView.tsx:46]: Why did you rename this? We can call it isAutenticated.

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments PTAL.

File: react/features/settings/actions.web.ts (hunk 7 above)
Comment (Calinteodor) [reply react/features/settings/actions.web.ts]: I moved inside onLogout function.

File: react/features/authentication/middleware.ts [hunk 9]
```diff
@@ -305,3 +279,27 @@ function _handleLogin({ dispatch, getState }: IStore) {
...
+    const { conference } = state['features/base/conference'];
+
+    const config = state['features/base/config'];
+    const logoutUrl = config.tokenLogoutUrl;
+
+    if (!conference) {
+        return;
+    }
+
+    if (logoutUrl) {
```
Comment (damencho) [inline react/features/authentication/middleware.ts:302]: This changes the behavior we had, now logout works only with logout URL. 
If you have configured your deployment with authentication with username and password and anonymous domain those that are authenticated can do a logout which sends an iq to jicofo and then hangups redirecting to welcomepage.

Comment (saghul) [inline react/features/authentication/middleware.ts:302]: This doesn't take into consideration the non token auth.

For non-token auth all we need to do is the moderator logout and hangup.

File: react/features/settings/actions.web.ts [hunk 10]
```diff
@@ -28,15 +33,40 @@ import {
...
+                    if (browser.isElectron()) {
+                        const url = appendURLHashParam(logoutUrl, 'electron', 'true');
+
+                        window.open(url, '_blank');
+                        dispatch(hangup(true));
+                    } else {
+                        window.location.href = logoutUrl;
+                    }
+
+                    conference?.room.moderator.logout(() => dispatch(hangup(true)));
```
Comment (saghul) [inline react/features/settings/actions.web.ts:65]: This is for the non-token case.

**Review by saghul — COMMENTED** [review]
Almost there!

File: react/features/settings/actions.web.ts (hunk 10 above)
Comment (damencho) [inline react/features/settings/actions.web.ts:65]: The described behavior above needs this logout to happen if token auth url is not set. We send this logout iq to jicofo when we are using username and password login.

File: lang/main.json [hunk 11]
```diff
@@ -250,7 +250,7 @@
     "dialog": {
         "Back": "Back",
         "Cancel": "Cancel",
-        "IamHost": "Log-in",
+        "IamHost": "Sign in",
```
Comment (Calinteodor) [inline lang/main.json:253]: @damencho @saghul I suggest we switch to sign in and sign out for action buttons because this is more consistent regarding the fact that we sign in via google, github or facebook.

Comment (saghul) [reply lang/main.json:253]: The terminology was discussed with Emil, please don't change it.

Comment (Calinteodor) [reply lang/main.json:253]: Ok. What should the action buttons on ProfileView read? Log-in Log-out?

Comment (saghul) [reply lang/main.json]: Yes please!

Comment (Calinteodor) [reply lang/main.json]: Perfect! Updated!

File: react/features/settings/actions.web.ts (hunk 10 above)
Comment (damencho) [reply react/features/settings/actions.web.ts:66]: Sorry but I still think this is changed.

In old we can execute dispatch(setTokenAuthUrlSuccess(false)) and after that do redirect or moderator.logout.

While here if we do dispatch(setTokenAuthUrlSuccess(false)) there is nothing after that.

File: react/features/settings/actions.native.ts [hunk 12]
```diff
@@ -1,10 +1,34 @@
...
+    return (dispatch: IStore['dispatch'], getState: IStore['getState']) => {
+        const state = getState();
+        const { conference } = state['features/base/conference'];
+
+        const config = state['features/base/config'];
+        const logoutUrl = config.tokenLogoutUrl;
+
+        dispatch(openDialog(LogoutDialog, {
+            onLogout() {
+                if (logoutUrl) {
```
Comment (saghul) [inline react/features/settings/actions.native.ts:25]: Add an extra check to see if token auth is enabled, don't just rely on a logout url being present.

Comment (Calinteodor) [reply react/features/settings/actions.native.ts]: Updated

File: react/features/settings/actions.native.ts [hunk 13]
```diff
@@ -1,10 +1,37 @@
...
+
+        dispatch(openDialog(LogoutDialog, {
+            onLogout() {
+                if (isTokenAuthEnabled(config)) {
+                    if (logoutUrl) {
+                        Linking.openURL(logoutUrl);
+                        dispatch(hangup(true));
+                    }
+
+                    conference?.room.moderator.logout(() => dispatch(hangup(true)));
```
Comment (saghul) [inline react/features/settings/actions.native.ts:32]: This line should go on an else, in canse token auth is NOT enabled.

File: react/features/settings/actions.native.ts [hunk 14]
```diff
@@ -1,10 +1,37 @@
...
+
+        const config = state['features/base/config'];
+        const logoutUrl = config.tokenLogoutUrl;
+
+        dispatch(openDialog(LogoutDialog, {
+            onLogout() {
+                if (isTokenAuthEnabled(config)) {
+                    if (logoutUrl) {
+                        Linking.openURL(logoutUrl);
+                        dispatch(hangup(true));
```
Comment (saghul) [inline react/features/settings/actions.native.ts:29]: What should happen if you click logout but there is no URL? I'd say we just hangup, so we'd need to take this line out of the if.

File: react/features/settings/actions.web.ts [hunk 15]
```diff
@@ -28,15 +33,45 @@ import {
...
+                    const url = appendURLHashParam(logoutUrl, 'electron', 'true');
+
+                    window.open(url, '_blank');
+                    dispatch(hangup(true));
+                } else {
+                    if (logoutUrl) {
+                        window.location.href = logoutUrl;
+                    }
+
+                    conference?.room.moderator.logout(() => dispatch(hangup(true)));
```
Comment (saghul) [inline react/features/settings/actions.web.ts:70]: This only runs if the auth mode is NOT token.

File: react/features/settings/actions.web.ts [hunk 16]
```diff
@@ -28,15 +33,45 @@ import {
...
+
+        const config = state['features/base/config'];
+        const logoutUrl = config.tokenLogoutUrl;
+
+        const { conference } = state['features/base/conference'];
+        const { jwt } = state['features/base/jwt'];
+
+        dispatch(openDialog(LogoutDialog, {
+            onLogout() {
+                if (isTokenAuthEnabled(config) && config.tokenAuthUrlAutoRedirect && jwt) {
```
Comment (saghul) [inline react/features/settings/actions.web.ts:54]: I think you need to separate these 2 checks to make things clearer.
